cpp/cmake_modules/ThirdpartyToolchain.cmake

```diff
@@ -4485,115 +4485,135 @@ target_include_directories(arrow::hadoop INTERFACE "${HADOOP_HOME}/include")
+  set(LZ4_HOME
+      ${LZ4_ROOT}
+      CACHE BOOL "" FORCE)
+  set(LZ4_LIBRARY
+      LZ4::lz4
+      CACHE STRING "" FORCE)
+  set(LZ4_STATIC_LIB
+      LZ4::lz4
+      CACHE STRING "" FORCE)
 
-  get_target_property(ORC_PROTOBUF_ROOT ${ARROW_PROTOBUF_LIBPROTOBUF}
+  get_target_property(PROTOBUF_INCLUDE_DIR ${ARROW_PROTOBUF_LIBPROTOBUF}
                       INTERFACE_INCLUDE_DIRECTORIES)
-  get_filename_component(ORC_PROTOBUF_ROOT "${ORC_PROTOBUF_ROOT}" DIRECTORY)
-
-  get_target_property(ORC_SNAPPY_INCLUDE_DIR ${Snappy_TARGET}
+  get_filename_component(Protobuf_ROOT "${PROTOBUF_INCLUDE_DIR}" DIRECTORY)
+  set(PROTOBUF_HOME
+      ${Protobuf_ROOT}
+      CACHE BOOL "" FORCE)
+  target_include_directories(${ARROW_PROTOBUF_LIBPROTOC}
+                             INTERFACE "${PROTOBUF_INCLUDE_DIR}")
+  set(PROTOBUF_EXECUTABLE $<TARGET_FILE:${ARROW_PROTOBUF_PROTOC}>)
+  set(PROTOBUF_LIBRARY $<TARGET_FILE:${ARROW_PROTOBUF_LIBPROTOBUF}>)
+  set(PROTOC_LIBRARY $<TARGET_FILE:${ARROW_PROTOBUF_LIBPROTOC}>)
+
+  get_target_property(SNAPPY_INCLUDE_DIR ${Snappy_TARGET} INTERFACE_INCLUDE_DIRECTORIES)
+  get_filename_component(Snappy_ROOT "${SNAPPY_INCLUDE_DIR}" DIRECTORY)
+  set(SNAPPY_HOME
+      ${Snappy_ROOT}
+      CACHE BOOL "" FORCE)
+  set(SNAPPY_LIBRARY
+      ${Snappy_TARGET}
+      CACHE STRING "" FORCE)
+
+  get_target_property(ZLIB_INCLUDE_DIR ZLIB::ZLIB INTERFACE_INCLUDE_DIRECTORIES)
+  get_filename_component(ZLIB_ROOT "${ZLIB_INCLUDE_DIR}" DIRECTORY)
+  set(ZLIB_HOME
+      ${ZLIB_ROOT}
+      CACHE BOOL "" FORCE)
+  set(ZLIB_LIBRARY
+      ZLIB::ZLIB
+      CACHE STRING "" FORCE)
+  get_target_property(ZLIB_TYPE ZLIB::ZLIB TYPE)
+  if(ZLIB_TYPE STREQUAL "STATIC_LIBRARY")
+    set(ZLIB_STATIC_LIB
+        ZLIB::ZLIB
+        CACHE STRING "" FORCE)
+  endif()
+
+  get_target_property(ZSTD_INCLUDE_DIR ${ARROW_ZSTD_LIBZSTD}
                       INTERFACE_INCLUDE_DIRECTORIES)
-  get_filename_component(ORC_SNAPPY_ROOT "${ORC_SNAPPY_INCLUDE_DIR}" DIRECTORY)
-
-  get_target_property(ORC_LZ4_ROOT LZ4::lz4 INTERFACE_INCLUDE_DIRECTORIES)
-  get_filename_component(ORC_LZ4_ROOT "${ORC_LZ4_ROOT}" DIRECTORY)
+  get_filename_component(ZSTD_ROOT "${ZSTD_INCLUDE_DIR}" DIRECTORY)
+  set(ZSTD_HOME
+      ${ZSTD_ROOT}
+      CACHE BOOL "" FORCE)
+  set(ZSTD_LIBRARY $<TARGET_FILE:${ARROW_ZSTD_LIBZSTD}>)
+  set(ZSTD_STATIC_LIBRARY $<TARGET_FILE:${ARROW_ZSTD_LIBZSTD}>)
+
+  set(BUILD_CPP_TESTS
+      OFF
+      CACHE BOOL "" FORCE)
+  set(BUILD_JAVA
+      OFF
+      CACHE BOOL "" FORCE)
+  set(BUILD_LIBHDFSPP
+      OFF
+      CACHE BOOL "" FORCE)
+  set(BUILD_TOOLS
+      OFF
+      CACHE BOOL "" FORCE)
+  set(INSTALL_VENDORED_LIBS
+      OFF
+      CACHE BOOL "" FORCE)
+  set(STOP_BUILD_ON_WARNING
+      OFF
+      CACHE BOOL "" FORCE)
 
-  get_target_property(ORC_ZSTD_ROOT ${ARROW_ZSTD_LIBZSTD} INTERFACE_INCLUDE_DIRECTORIES)
-  get_filename_component(ORC_ZSTD_ROOT "${ORC_ZSTD_ROOT}" DIRECTORY)
+  # TODO: This should be fixed in upstream.
```

Comment:
Can we remove this after upgrading to v2.0.2 which should the fix? If yes, perhaps add a comment to we won't forget.